Default dt is 0

sawyerbfuller opened the pull request:
Changes the new default to be ``dt=0`` (continuous-time), rather than ``dt=None`` (timebase unspecified) for state space, transfer function, and input output systems. This follows matlab convention as discussed in issue #422
* new default is ``dt=0`` in these three modules (edit: setting now is in a single location: ``defaults['control.default_dt']``)
* to make interconnection behavior match what is described in [conventions](https://python-control.readthedocs.io/en/0.8.3/conventions.html#discrete-time-systems), a new function: ``lti.common_timebase(dt1, dt2)`` was introduced. It performs logic to correctly combine timebases of combined systems.  The new function now correctly allows discrete-time systems with unspecified sampling time (``dt=True``) to be combined with systems with specified sampling time (``dt!=0``, ``dt!=None``, eg ``dt=0.1``). 
* new functions ``_isstaticgain`` (edit: now is a method ``is_static_gain``) were added to detect upon initialization whether a state space or transfer function is a static gain. If so, ``dt=None`` for such systems so they may be combined with either discrete- or continuous-time systems. 
* I needed to take dt out of the arguments list in ``iosys`` init functions because library-wide defaults are not available when modules are first imported, which is when default values in function declarations are evaluated. My solution was to move the dt keyword in the ``**kwargs`` dict.  
* fixes to unit tests
* ``use_legacy_defaults('0.8.3')`` reverts to old default behavior of ``dt=None``

sawyerbfuller commented:
Let me know what you think. Still to do: update documentation in modules and in conventions.rst

sawyerbfuller commented:
some remarks:
* most of the unit tests worked as-is with the change, except for ones that implicitly assumed dt=None during the tests. After making that assumption explicit on such those systems, pretty much everything worked
* most of the work was in implementing a test for static gain in StateSpace and TransferFunction (maybe one is needed in iosys?) 
* and in standardizing the use of ``lti.common_timebase`` everywhere in the software when interconnecting systems
* I tested it on a few pages of code full of discrete and cont time systems and everything worked unchanged

murrayrm commented:
For the defaults, it seems like most people are going use the same default for transfer functions, state space systems, and I/O systems.  Right now they have to set three defaults.  Does it makes sense to collapse this to just one configuration variable?  Or perhaps allow both: a "global" `default_dt` with individual overrides if you want them.

Also: since this will nominally break code, I propose putting this into v0.9 (versus 0.8.4, which we are close to releasing).

Will try to take a closer look at the code itself in a bit.

sawyerbfuller commented:
Hi Richard, this all makes sense and sounds reasonable. There was some reason (I think it was mostly expedience) for why I didn’t just use a a global default dt - I think I had decided that something about the global defaults hadn’t been completely implemented yet, but let me take a look, I agree that would be the right way to do it if possible.

bnavigator commented on control/lti.py:
```diff
@@ -157,48 +157,31 @@ def timebase(sys, strict=True):
     return sys.dt
 
-# Check to see if two timebases are equal
-def timebaseEqual(sys1, sys2):
```
Removing this function here and from `__all__` may break existing code. Maybe introduce a `DeprecationWarning` between 0.9 and 0.10 or 0.9.1?

sawyerbfuller replied:
Done.

bnavigator commented on control/tests/discrete_test.py:
```diff
@@ -123,41 +130,18 @@ def test_timebase_conversions(self):
-            raise ValueError("incompatible operation allowed")
-        except ValueError:
-            pass
+        self.assertRaises(ValueError, tf2.__mul__, tf3)
+        self.assertRaises(ValueError, tf2.__mul__, tf4)
+        self.assertRaises(ValueError, tf2.__add__, tf3)
+        self.assertRaises(ValueError, tf2.__add__, tf4)
+        self.assertRaises(ValueError, tf2.feedback, tf3)
+        self.assertRaises(ValueError, tf2.feedback, tf4)
```
People have been fiddling with `__rmul__` and co. It may be better to use

```python
with self.assertRaises(ValueError):
    tf2*tf3
```

sawyerbfuller replied:
Thanks for the kinda words! 

I will give that test  formulation a try and leave in that function with a deprecation warning next time I’m at my computer.

sawyerbfuller commented:
OK made a few suggested changes: 
* default.dt is now package-wide. change it with ``set_defaults('control', default_dt=0)``
* re-incorporated timebaseEqual with deprecation warning
* improved a few unit tests

bnavigator commented on .vscode/settings.json:
```diff
@@ -0,0 +1,3 @@
+{
+    "restructuredtext.confPath": "${workspaceFolder}/doc"
+}
```
Remove this directory and file, please.

sawyerbfuller replied:
good catch, sorry, done.

sawyerbfuller commented:
latest commit creates a new, potentially-useful method, ``is_static_gain()`` for StateSpace and TransferFunction systems. refactored their __init__ methods to use it now rather than as an internal method. 
* during refactor, added the ability to __init__ using a dt keyword, as suggested in issue #419. If dt is received as both a positional and a keyword argument, the user receives a warning and the positional argument is used.

sawyerbfuller commented:
I think this should also resolve #399

bnavigator commented:
Did someone check to merge your PRs with @samlaf's? They seem to work on overlapping funcitonality...

sawyerbfuller commented:
> Did someone check to merge your PRs with @samlaf's? They seem to work on overlapping functionality.

@bnavigator good to check. A brief glance through suggests that they are are basically operating on different aspects of the iosys functionality, and I don't anticipate any merge problems, except for the occasional empty line. The devil is in the details, of course. I don't know git will enough to do a test merge. If any issues crop up you can let me know and I will be happy to take a look at it.

bnavigator commented:
I made a test merge in https://github.com/bnavigator/python-control/pull/1 and reactivated TravisCI for my fork. Tests look good so far, there were merge conflicts in both merges, because the edits were in lines next to each other. Please check, if I resolved correctly.

_Originally posted by @bnavigator in https://github.com/python-control/python-control/pull/433#issuecomment-663235302_

I postet the comment in the wrong thread. Meant to post it here, so that @samlaf gets notified too. Alas, another mention.

bnavigator commented on control/statesp.py:
```diff
@@ -207,12 +209,33 @@ def __init__(self, *args, **kw):
+            if 'dt' in kwargs:
+                warn('received multiple dt arguments, using positional arg dt=%s'%dt)
+        elif len(args) == 1:
+            try:
+                dt = args[0].dt
+            except NameError:
+                if self.is_static_gain(): 
+                    dt = None
+                else:
+                    dt = config.defaults['control.default_dt']
```
Does it ever happen that the passed `StateSpace` object (no other type allowed) has no `dt` argument? If so, it will rise an `AttributeError` instead of `NameError`

bnavigator replied:
For a unit test covering this, see test_copy_constructor_nodt() in https://github.com/python-control/python-control/pull/438/commits/ee21d548d10f93de96210ef4d18adeadf03ca3c4 (#438)

sawyerbfuller commented:
Since you bring it up, I suspect this is leftover from earlier code. The
lti init  method always sets dt, and it is called any time a tf or ss is
instantiated. So there will always be a dt.  I think the best thing is to
eliminate the try...except clause. Sawyer

On Tue, Aug 4, 2020 at 5:39 AM Ben Greiner <notifications@github.com> wrote:

> *@bnavigator* commented on this pull request.
> ------------------------------
>
> In control/statesp.py
> <https://github.com/python-control/python-control/pull/431#discussion_r465018602>
> :
>
> > +            try:
> +                dt = args[0].dt
> +            except NameError:
> +                if self.is_static_gain():
> +                    dt = None
> +                else:
> +                    dt = config.defaults['control.default_dt']
>
> Does it ever happen that the passed StateSpace object (no other type
> allowed) has no dt argument? If so, it will rise an AttributeError
> instead of NameError
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/python-control/python-control/pull/431#pullrequestreview-460771710>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AN74SSOAPNC7FPBBM3RFHS3R676OFANCNFSM4O7ABJRQ>
> .
>
-- 
Sawyer Fuller
Assistant Professor of Mechanical Engineering
Adjunct, Paul G Allen School of Computer Science
University of Washington
http://faculty.washington.edu/~minster/

(Typed with my thumbs)

murrayrm commented:
That code might have been there to enable [duck typing](https://en.wikipedia.org/wiki/Duck_typing).

bnavigator commented:
Ah, now I see you did not introduce these lines but just moved them. Anyway, they are not covered in the code version and would not work, because:
- no other object than `StateSpace` can be passed to the constructor and reach this code. It fails a few lines before (which is the duck typing experience if I understand @murrayrm's comment correctly)
- if the `dt` attribute is not forcefully removed from the `StateSpace` object it will always be there.
- if it is really not present, the try clause will raise an `AttributeError` instead of `NameError`.

sawyerbfuller commented:
Following richard’s comment, I agree duck typing is probably the right way
to do this. Which would entail leaving that code as it is and only raising
an error farther above if the object that is passed doesn’t have an A, B,
C, and D associated with it. What do you think?

The easiest way would be to just do something like:

self.A = args[0].A
self.B = ...

And just let the attributeError raise itself.

A more user-friendly version would raise an error with a statement to the
effect “argument must have A, B, C, and D attributes to be converted into a
state-space system”.

I would favor the latter, what do you think?

Happy to make a PR if no objections. S


On Tue, Aug 4, 2020 at 10:43 AM Ben Greiner <notifications@github.com>
wrote:

> Ah, now I see you did not introduce these lines but just moved them.
> Anyway, they are not covered in the code version and would not work,
> because:
>
>    - no other object than StateSpace can be passed to the constructor and
>    reach this code. It fails a few lines before (which is the duck typing
>    experience if I understand @murrayrm <https://github.com/murrayrm>'s
>    comment correctly)
>    - if the dt attribute is not forcefully removed from the StateSpace
>    object it will always be there.
>    - if it is really not present, the try clause will raise an
>    AttributeError instead of NameError.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/python-control/python-control/pull/431#issuecomment-668733603>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AN74SSI46NDHL4UJDEG6F4DR7BCBPANCNFSM4O7ABJRQ>
> .
>
-- 
Sawyer Fuller
Assistant Professor of Mechanical Engineering
Adjunct, Paul G Allen School of Computer Science
University of Washington
http://faculty.washington.edu/~minster/

(Typed with my thumbs)

bnavigator commented:
> The easiest way would be to just do something like:
>     self.A = args[0].A
>     self.B = ...

I would at least call `StateSpace(args[0].A, args[0].B, ...)` to make sure the attributes convert to valid matrices.

`ss()` should be changed accordingly.

sawyerbfuller commented:
Thinking about adding duck typing a little more, and where to add the relevant code, I can't figure out why ``StateSpace.__init__`` and ``ss`` don't share the same code. 

I understand why it might make sense to keep core functionality and the Matlab layer different, but right now it looks like it would be possible for them to just share the same code (by having ``ss`` just call ``StateSpace(*args, **kwargs)``) while also maintaining backward compatibility and a superset of their combined current capability, as well as duck typing, and less bugs. Currently, they have subtle differences and duplicate functionality in different chunks of code, but their call signatures are basically identical. (same holds for ``TransferFunction`` and ``tf``). The only reason I could think of for keeping them different is if there is a plan for the ``StateSpace`` constructor to change or add in new functionality in the future? 

Background: 
``ss`` 
    * automatically converts a ``TransferFunction`` into a ``StateSpace`` system if it is passed as an argument
    * but can't accommodate extra keywords like ``remove_useless_states`` if a it receives a system as its argument
``StateSpace``: 
    * raises an error if given a ``TransferFunction`` system -- does not auto-convert
* neither do duck typing - if they are given an object with A, B, C, and D attributes that is not a ``StateSpace`` system, an error is raised. not very pythonic
* it is similar for ``TransferFunction`` and ``tf``

murrayrm commented:
I don't remember the history of why `ss` doesn't just call `StateSpace`.  The one thing I can think of is that you can't really "create" a state space system from a transfer function.  You can *realize* a transfer function as a state space system, but of course the realization is not unique => you need some additional information to specify the realization.  Putting this in the `StateSpace` constructor seems odd.  It really belongs in a function with a name like `tf_realize` or something similar (right now I guess it is in `tf2ss`).

Any of these variations probably belongs in a separate PR...

sawyerbfuller commented:
Makes sense, I'll leave them as separate functions. Maybe ``ss`` auto-realizes (converts) to duplicate matlab functionality. Maybe if/when this current PR goes in, then I'll make a new PR that adds some comments to the code to explain this rationale and add in duck typing to ``StateSpace``.

bnavigator commented:
Superseded by #490
